### PR TITLE
DivRem 64by32 and 32by16 refactor in Decimal.DecCalc

### PR DESCRIPTION
**Pull request by lilinus**
Tried to make the code easier to read with all the DivRem stuff going on. It will also be much easier to leverage intrinsics in the helper methods (such as `X86Base.DivRem`)  when they can be used properly.

I don't know how to compare benchmarks thought, to see there are no regressions in performance.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Decimal.DecCalc.cs [hunk 1]
```diff
@@ -366,10 +364,9 @@ private static uint Div96By64(ref Buf12 bufNum, ulong den)
                     // Result is zero.  Entire dividend is remainder.
                     //
                     return 0;
-
-                // TODO: https://github.com/dotnet/runtime/issues/5213
-                quo = (uint)(num64 / denHigh32);
-                num = bufNum.U0 | ((num64 - quo * denHigh32) << 32); // remainder
+;
```
Comment (lilinus) [inline src/libraries/System.Private.CoreLib/src/System/Decimal.DecCalc.cs:367]: Mistake, will delete this stray semicolon.

**Comment by danmoseley** [comment]
> I don't know how to compare benchmarks thought, to see there are no regressions in performance.

Have you looked at the docs in dotnet/performance? They explain how to test private changes like this. If there's not coverage, you can add a test there.

**Comment by tannergooding** [comment]
Same general comments that I gave on https://github.com/dotnet/runtime/pull/99212 apply here.

These really seems like some basic JIT optimizations we're missing and not something we should be adding hardcoded helpers around. Cases of `ulong / uint` and similar are areas that the JIT should be able to recognize and optimize itself, allowing this to light up anywhere its used and not require customized logic that only works for `System.Decimal`.

This would be particularly relevant for `Arm64` which doesn't have specialized instructions allowing `uint128 / uint64` or similar.

**Comment by lilinus** [comment]
> These really seems like some basic JIT optimizations we're missing and not something we should be adding hardcoded helpers around. Cases of `ulong / uint` and similar are areas that the JIT should be able to recognize and optimize itself, allowing this to light up anywhere its used and not require customized logic that only works for `System.Decimal`.

I agree with the missing JIT optimizations. The purpose of this PR was not to optimize, rather refactor a bit without impacting performance negatively.

Main purpose of this PR is:
- Make code more readable and easier to follow / maintain using the helpers. IMO it's just easier to follow what is going on with the divide operators with carries with these introduced helpers.
- Homogenous DivRem methods already exists in `Math`, but these non-homogenous also (hopefully) conveys what the input and output of the operations better than before.

**Pull request closed without merging**